### ISSUE TITLE
Tree Path Dependent expected value

### ISSUE THREAD

**Issue by jacekblaz** [shap/shap#1082]
Hi folks,

I stuck on this topic for couple days and seems I need your help to understand what is expected value in TreeExplainer when feature_perturbation = tree_path_dependent, precisely what is expected value in this code?

 _I use SHAP 0.35, xgboost._

```
explainer = shap.TreeExplainer(model=model, feature_perturbation='tree_path_dependent', model_output='raw') 
expected_value = explainer.expected_value
```

I know that if I use feature_perturbation = interventional then expected_value is just mean log odds from predictions:
```
explainer = shap.TreeExplainer(model=model, feature_perturbation='interventional', data= background_data)

#almost true
explainer.expected_value == model.predict(background_data, output_margin=True).mean()
```

Can you tell me what is it and how is obtained expected_value in the first code snippet?

**Comment by Gimba** [comment]
I have the same question. Did you find an answer yet?

**Comment by jacekblaz** [comment]
@Gimba I was looking into source code for couple hours and couldn't figure it out. 
Posted on CrossValidated and Data Science stack exchanges without any respons. 
I have to deliver soon so decided to use background data to keep xai under my control. I will probably dig once again when have time to do that.

**Comment by alexcoca** [comment]
Hey @Gimba, @jacekblaz,

Note that although a background dataset is not provided, the expected values is computed using the node cover information, stored at each node. The computation proceeds recursively, starting at the root. The contribution of a note to the expected value of the tree is a function of the expected values of the children and is computed as follows:

c_j = (c_{r(j)}*r_{r(j)} + c_{l(j)}*r_{l(j)}})/r_j

where j denotes the node index, c_j denotes the node expected value, r_j is the cover of the jth node and r(j) and l(j) represent the indices of the right and left children, respectively. The expected value used by the tree is simply c_{root}. Note that for tree ensamles, the expected values of the ensamble members is weighted according to the tree weight and the weighted expected values of all trees are summed to obtain a single value.

The cover depends on the objective function and the model chosen. For example, in a gradient boosted tree trained with squared loss objective, r_j is simply the number of training examples flowing through j. For an arbitrary objective, this is the sum of the hessian of the loss function evaluated at each point flowing through j.

I hope this helps. I could also point you out to the relevant parts of the `xgboost` codebase that I read to draw the above conclusions if you want to cross-check my conclusions. Happy to respond on CrossValidated DataScience stack exchange if you want.

**Comment by zyxue** [comment]
@alexcoca , do you have insight (or intuition) into why `tree_path_dependent` is a valid way of estimating `E[f(x)|x_S]`, esp. when S is not the empty set?  I haven't found any discussion in Algorithm 1 of https://arxiv.org/pdf/1905.04610.pdf on how such a way of estimation comes to be.

Note when S is the empty set,  `E[f(x)|x_S]` is called expected_value in code, essentially the average prediction of the model on the training data in the context of `tree_path_dependent` estimation.

**Comment by dwolfeu** [comment]
https://stackoverflow.com/questions/77126001/calculation-of-expected-value-in-shap-explanations-of-xgboost-classifier
